cmaes/_cmawm.py

```diff
@@ -0,0 +1,649 @@
+
+        .. code::
+
+            import numpy as np
+            from cmaes import CMAwM
+
+            def ellipsoid_onemax(x, n_zdim):
+                n = len(x)
+                n_rdim = n - n_zdim
+                ellipsoid = sum([(1000 ** (i / (n_rdim - 1)) * x[i]) ** 2 for i in range(n_rdim)])
+                onemax = n_zdim - (0. < x[(n - n_zdim):]).sum()
+                return ellipsoid + 10 * onemax
+
+            binary_dim, continuous_dim = 10, 10
+            dim = binary_dim + continuous_dim
+            bounds = np.concatenate(
+                [
+                    np.tile([0, 1], (binary_dim, 1)),
+                    np.tile([-np.inf, np.inf], (continuous_dim, 1)),
+                ]
+            )
+            steps = np.concatenate([np.ones(binary_dim), np.zeros(continuous_dim)])
+            optimizer = CMAwM(mean=np.zeros(dim), sigma=2.0, bounds=bounds, steps=steps)
+
+            evals = 0
+            while True:
+                solutions = []
+                for _ in range(optimizer.population_size):
+                    x_for_eval, x_for_tell = optimizer.ask()
+                    value = ellipsoid_onemax(x_for_eval, binary_dim)
+                    evals += 1
+                    solutions.append((x_for_tell, value))
+                optimizer.tell(solutions)
+
+                if optimizer.should_stop():
+                    break
+
+    Args:
+
+        mean:
+            Initial mean vector of multi-variate gaussian distributions.
+
+        sigma:
+            Initial standard deviation of covariance matrix.
+
+        bounds:
+            Lower and upper domain boundaries for each parameter.
+
+        steps:
+            Each value represents a step of discretization for each dimension.
+            Zero (or negative value) means a continuous space.
+
+        n_max_resampling:
+            A maximum number of resampling parameters (default: 100).
+            If all sampled parameters are infeasible, the last sampled one
+            will be clipped with lower and upper bounds.
+
+        seed:
+            A seed number (optional).
+
+        population_size:
+            A population size (optional).
+
+        cov:
+            A covariance matrix (optional).
+
+        margin:
+            A margin parameter (optional).
+    """
+
+    # Paper: https://arxiv.org/abs/2205.13482
+
+    def __init__(
+        self,
+        mean: np.ndarray,
+        sigma: float,
+        bounds: np.ndarray,
+        steps: np.ndarray,
+        n_max_resampling: int = 100,
+        seed: Optional[int] = None,
+        population_size: Optional[int] = None,
+        cov: Optional[np.ndarray] = None,
+        margin: Optional[float] = None,
+    ):
+        assert len(bounds) == len(steps), "bounds and steps must be the same length"
+        assert not np.isnan(steps).any(), "steps should not include NaN"
+
+        # split discrete space and continuous space
+        self._discrete_idx = np.where(steps > 0)[0]
```

Comment:
Should we add an assertion to ensure the existence of discrete parameters?

Reply:
Thanks for pointing it out! I added that validation. (https://github.com/CyberAgentAILab/cmaes/pull/121/commits/1ef697199592f2e2433fa3648c95ff3d24afd3b5)